File: instrumentation/play-2.7/src/main/scala/play/core/server/Server_Instrumentation.scala
/*
 *
 *  * Copyright 2020 New Relic Corporation. All rights reserved.
 *  * SPDX-License-Identifier: Apache-2.0
 *
 */

package play.core.server

import com.newrelic.api.agent.weaver.Weaver
import com.newrelic.api.agent.weaver.scala.{ScalaMatchType, ScalaWeave}
import com.newrelic.api.agent.{NewRelic, Trace}
import com.nr.agent.instrumentation.play27.PlayUtil
import play.api.Application
import play.api.mvc.{Handler, RequestHeader}

import scala.util.Try;

@ScalaWeave(`type` = ScalaMatchType.Object, `originalName` = "play.core.server.Server")
class Server_Instrumentation {

  @Trace(dispatcher = true)
  private[server] def getHandlerFor(request: RequestHeader, tryApp: Try[Application]): (RequestHeader, Handler) = {
    var result: (RequestHeader, Handler) = Weaver.callOriginal()

    // In order to correctly track async work from here we need to stash a token into the request.
    val token = NewRelic.getAgent.getTransaction.getToken
    result = PlayUtil.appendToken(result, token)

    val tracer = NewRelic.getAgent.getTracedMethod
    if (tracer != null) {
      tracer.setMetricName("Play2Routing");
    }

    result
  }

}

File: instrumentation/jetty-12/src/main/java/org/eclipse/jetty/server/Server_Instrumentation.java
/*
 *
 *  * Copyright 2023 New Relic Corporation. All rights reserved.
 *  * SPDX-License-Identifier: Apache-2.0
 *
 */

package org.eclipse.jetty.server;

import com.newrelic.agent.bridge.AgentBridge;
import com.newrelic.api.agent.weaver.MatchType;
import com.newrelic.api.agent.weaver.Weave;
import com.newrelic.api.agent.weaver.Weaver;
import com.nr.agent.instrumentation.jetty12.JettyRequest;
import com.nr.agent.instrumentation.jetty12.JettyResponse;
import com.nr.agent.instrumentation.jetty12.JettySampler;
import org.eclipse.jetty.util.Callback;
import org.eclipse.jetty.util.thread.ThreadPool;

import java.util.concurrent.TimeUnit;

@Weave(type = MatchType.ExactClass, originalName = "org.eclipse.jetty.server.Server")
public abstract class Server_Instrumentation {

    protected void doStart() {
        AgentBridge.privateApi.addSampler(new JettySampler(this), 1, TimeUnit.MINUTES);
        Weaver.callOriginal();
    }

    public boolean handle(Request request, Response response, Callback callback) {
        boolean isStarted = AgentBridge.getAgent().getTransaction().isStarted();
        // if there is a #ContextHolder, then this is not embedded Jetty, so the transaction should start there
        boolean startTransaction = request != null && !isStarted;

        if (startTransaction) {
            AgentBridge.getAgent().getTransaction(true)
                    .requestInitialized(new JettyRequest(request), new JettyResponse(response));
        }

        boolean result = false;
        try {
            result = Weaver.callOriginal();
        } finally {
            if (startTransaction) {
                AgentBridge.getAgent().getTransaction().requestDestroyed();
            }
        }

        return result;
    }

    public static String getVersion() {
        return Weaver.callOriginal();
    }

    public abstract Connector[] getConnectors();

    public abstract ThreadPool getThreadPool();
}
